### PR TITLE
The next major release.

### PR DESCRIPTION
**Pull request by aq1018**
Placeholder for discussing next major release.

The goal for next release:
- Cleanup code and simplify architecture.
- Reporters. Only 3 types now. `array`, `hash`, and `message`. They use formatters to format messages.
- Support custom Reporter with configuration.
- Formatters. Comes with `human` and `machine` flavors.
- Support custom Formatters.
- Support `model.errors.format(:machine).to_hash`.

### CODE REVIEW

**Comment by aq1018** [comment]
@krainboltgreene Do you mind taking a look at this branch? I did some pretty heavy refactoring and would like to know about your opinions on the convention / architecture, etc...

File: active_model-better_errors.gemspec [hunk 1]
```diff
@@ -4,30 +4,25 @@ lib = File.expand_path('../lib', __FILE__)
...
-  spec.summary       = "active_model-better_errors is a ActiveModel::Errors compatible library to help you customize the presentation of your error messages."
-  spec.homepage      = "http://github.com/aq1018/active_model-better_errors"
-  spec.license       = "MIT"
+Gem::Specification.new do |gem|
+  gem.name          = "active_model-better_errors"
+  gem.version       = ActiveModel::BetterErrors::VERSION
+  gem.authors       = ["Aaron Qian", "Byron Anderson"]
+  gem.email         = ["aq1018@gmail.com", "byronanderson32@gmail.com"]
+  gem.description   = "API consumable error messages with ActiveModel::Errors drop-in compatibility."
+  gem.summary       = "active_model-better_errors is a ActiveModel::Errors compatible library to help you customize the presentation of your error messages."
```
Comment (krainboltgreene) [inline active_model-better_errors.gemspec:13]: Just point `gem.description` to `gem.summary` and replace description with summary.

File: active_model-better_errors.gemspec [hunk 2]
```diff
@@ -4,30 +4,25 @@ lib = File.expand_path('../lib', __FILE__)
...
+  gem.homepage      = "http://github.com/aq1018/active_model-better_errors"
```
Comment (krainboltgreene) [inline active_model-better_errors.gemspec:14]: Should be `"https://aq1018.github.io/active_model-better_errors"` for the .io and then we can make a github page.

File: lib/active_model/better_errors/constants.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+module ActiveModel
+  module BetterErrors
+    REPORTER_TYPES = [:array, :hash, :message]
```
Comment (krainboltgreene) [inline lib/active_model/better_errors/constants.rb:6]: Might just be better to have this in `lib/active_model/better_errors.rb`

File: lib/active_model/better_errors/emulation.rb [hunk 4]
```diff
@@ -19,16 +19,23 @@ module Emulation
...
         base.class_eval do
-          extend Forwardable
-          def_delegators :error_collection, *MODEL_METHODS
-          def_delegators :message_reporter, *MESSAGE_REPORTER_METHODS
-          def_delegators :hash_reporter,    :to_hash
-          def_delegators :array_reporter,   :to_a
+          delegate(*MODEL_METHODS, to: :error_collection)
+          delegate(*MESSAGE_REPORTER_METHODS, to: :message_reporter)
+          delegate(*HASH_REPORTER_METHODS, to: :hash_reporter)
+          delegate(*ARRAY_REPORTER_METHODS, to: :array_reporter)
```
Comment (krainboltgreene) [inline lib/active_model/better_errors/emulation.rb:39]: This is pretty clever.

File: lib/active_model/better_errors/formatter/human.rb [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+          if base.class.respond_to?(:i18n_scope)
+            keys << :"#{base.class.i18n_scope}.errors.messages.#{type}"
+          end
+
+          keys << :"errors.attributes.#{attribute}.#{type}"
+          keys << :"errors.messages.#{type}"
+
+          keys.compact!
+          keys.flatten!
+          keys
```
Comment (krainboltgreene) [inline lib/active_model/better_errors/formatter/human.rb]: This is probably a method we should refactor into smaller pieces, or it's own object.

**Comment by krainboltgreene** [comment]
The changes are definitely for the better.

File: active_model-better_errors.gemspec (hunk 2 above)
Comment (aq1018) [inline active_model-better_errors.gemspec:14]: That's interesting. Do you want to make a static site for this?

File: lib/active_model/better_errors/constants.rb (hunk 3 above)
Comment (aq1018) [inline lib/active_model/better_errors/constants.rb:6]: Actually that won't work unfortunately. This constant needs to be declared head of time so other parts of the lib can use it to do some meta-programming. That's why this is extracted out of `lib/active_model/better_errors.rb` and then required as the first file.

File: lib/active_model/better_errors/formatter/human.rb (hunk 5 above)
Comment (aq1018) [inline lib/active_model/better_errors/formatter/human.rb]: Sounds good. I'll extract it into a `I18nKeyBuilder` class or something. BTW, this is already a cleaned up version of the original `ActiveModel::Errors` code. Rails code really needs some good cleanup.

**Pull request closed without merging**